### PR TITLE
.Net: Add support for making open api operation metadata and extension metadata available at function invocation time

### PR DESCRIPTION
**Pull request by westey-m**
### Motivation and Context

This change is to support scenarios where a consumer needs to use metadata about an operation to change behavior at function invocation time.

E.g. OpenAI has defined a [consequential flag](https://platform.openai.com/docs/actions/getting-started/consequential-flag) that can be added to open api operation schemas in order to indicate that the operation will have consequences if invoked. It may create a transaction in a database. Compare that to an operation that just retrieves data and has no additional consequences.

In some cases a consumer may want to use this flag as an input when deciding on whether to confirm an function invocations with a user before invocation. The LLM may have instructed the consumer to invoke the function, but since there is risk involved, the consumer may want to get user confirmation first.

Related Issues:
https://github.com/microsoft/semantic-kernel/issues/5724
https://github.com/microsoft/semantic-kernel/issues/5506

### Description

1. Added a readonly dictionary base class to KernelFunctionMetadata to allow adding any unstructured metadata to this dictionary in addition to the named properties already defined on KernelFunctionMetadata.
2. The base class approach was chosen because many other dictionary classes already exist for metadata and if we wanted to extend them with named properties a similar approach can be used without causing breaking changes.
3. When creating a plugin from an OpenAPI schema, properties for id, path and method are added to the metadata dictionary as well as an operation-extensions property, who's value is a dictionary and contains any x-* properties defined in the OpenAPI schema.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        // This HTTP client is optional. SK will fallback to a default internal one if omitted.
+        using HttpClient httpClient = new();
+
+        // Create a sample OpenAPI schema that calls the github versions api, and has an operation extension property.
+        // The x-openai-isConsequential property is the operation extension property.
+        var schema =
+            "{" +
+            "  \"openapi\": \"3.0.1\"," +
+            "  \"info\": {" +
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:38]: With recent upgrades of C# language version in repo, you should be able to use raw string literals, which should simplify usage and representation of JSON in code. For example:
https://github.com/microsoft/semantic-kernel/blob/c72080d7da6baa539eab515fc065730d87d191e2/dotnet/src/Connectors/Connectors.UnitTests/OpenAI/ChatCompletion/OpenAIChatCompletionServiceTests.cs#L352-L356

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            "          }" +
+            "        }" +
+            "      }" +
+            "    }" +
+            "  }" +
+            "}";
+        var schemaStream = new MemoryStream();
+        WriteStringToStream(schemaStream, schema);
+
+        // Import an Open AI plugin from a stream.
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:64]:
```suggestion
        // Import an OpenAPI plugin from a stream.
```

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        else
+        {
+            // Invoke the function and output the result.
+            var functionResult = await kernel.InvokeAsync(function, new KernelArguments());
+            var result = functionResult.GetValue<RestApiOperationResponse>();
+            WriteLine($"Function execution result: {result?.Content}");
+        }
+    }
+
+    public Example23_OpenAPIPlugin(ITestOutputHelper output) : base(output)
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:90]: nit: This constructor can be converted to primary constructor.

File: dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs [hunk 4]
```diff
@@ -200,6 +201,18 @@ public static class OpenApiKernelExtensions
...
+    private const string OperationExtensionsIdKey = "id";
+
+    /// <summary>The metadata property bag key to use when storing the path of an operation.</summary>
+    private const string OperationExtensionsPathKey = "path";
+
+    /// <summary>The metadata property bag key to use when storing the method of an operation.</summary>
+    private const string OperationExtensionsMethodKey = "method";
+
+    /// <summary>The metadata property bag key to use for the list of extension values provided in the swagger file at the operation level.</summary>
+    private const string OperationExtensionsMetadataKey = "operation-extensions";
```
Comment (dmytrostruk) [inline dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs:214]: Looks like `operation-extensions` key is used in example as well. Should we consider making this const public, so users will be able to re-use it? If we decide to convert them to public, we can think if we should keep them in `Extensions` class or more generic place.

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs [hunk 5]
```diff
@@ -19,12 +20,16 @@ public sealed class KernelFunctionMetadata
...
+    /// <summary>A static empty dictionary to default to when none is provided.</summary>
+    private static readonly Dictionary<string, object?> s_emptyDictionary = new();
 
     /// <summary>Initializes the <see cref="KernelFunctionMetadata"/> for a function with the specified name.</summary>
     /// <param name="name">The name of the function.</param>
+    /// <param name="propertyBag">Optional metadata in addition to the named properties already available on this class.</param>
     /// <exception cref="ArgumentNullException">The <paramref name="name"/> was null.</exception>
     /// <exception cref="ArgumentException">An invalid name was supplied.</exception>
-    public KernelFunctionMetadata(string name)
+    public KernelFunctionMetadata(string name, ReadOnlyDictionary<string, object?>? propertyBag = null)
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs:31]: Is there an option to add new constructor overload with `propertyBag` parameter instead of modifying existing constructor? In this case there will be no need in compatibility suppression file.

Also, should be name new parameter as `metadata` instead of `propertyBag`? I think `metadata` should work better, based on description of this parameter.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs [hunk 6]
```diff
@@ -25,15 +26,17 @@ public static class KernelFunctionFactory
...
     public static KernelFunction CreateFromMethod(
         Delegate method,
         string? functionName = null,
         string? description = null,
         IEnumerable<KernelParameterMetadata>? parameters = null,
         KernelReturnParameterMetadata? returnParameter = null,
-        ILoggerFactory? loggerFactory = null) =>
-        CreateFromMethod(method.Method, method.Target, functionName, description, parameters, returnParameter, loggerFactory);
+        ILoggerFactory? loggerFactory = null,
+        ReadOnlyDictionary<string, object?>? metadataPropertyBag = null) =>
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs:38]: Same comment as above - I would probably keep the naming simple and use just `metadata` or just `propertyBag`. 
Most probably it won't be possible to add method overload here, taking into account amount of optional parameters.

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs (hunk 1 above)
Comment (westey-m) [reply dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:38]: Oh yeah, that's much better. Will fix, thanks!

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs (hunk 3 above)
Comment (westey-m) [reply dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:90]: Makes sense.  Will fix.

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs (hunk 5 above)
Comment (dmytrostruk) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs:31]: Makes sense!

File: dotnet/src/SemanticKernel.Abstractions/CompatibilitySuppressions.xml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+<!-- https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids -->
+<Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <Suppression>
+    <DiagnosticId>CP0002</DiagnosticId>
+    <Target>M:Microsoft.SemanticKernel.KernelFunctionMetadata.#ctor(System.String)</Target>
+    <Left>lib/netstandard2.0/Microsoft.SemanticKernel.Abstractions.dll</Left>
+    <Right>lib/netstandard2.0/Microsoft.SemanticKernel.Abstractions.dll</Right>
+    <IsBaselineSuppression>true</IsBaselineSuppression>
+  </Suppression>
+</Suppressions>
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/CompatibilitySuppressions.xml:11]: As a general rule we never want to introduce breaking changes.

We've done this early on in the 1.0 cycle and for changes related to experimental features but let's try to avoid this.

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs (hunk 5 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs:31]: How about `additionalMetadata` or `extensionMetadata`

File: dotnet/src/SemanticKernel.Core/CompatibilitySuppressions.xml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    <IsBaselineSuppression>true</IsBaselineSuppression>
+  </Suppression>
+  <Suppression>
+    <DiagnosticId>CP0002</DiagnosticId>
+    <Target>M:Microsoft.SemanticKernel.KernelFunctionFactory.CreateFromMethod(System.Reflection.MethodInfo,System.Object,System.String,System.String,System.Collections.Generic.IEnumerable{Microsoft.SemanticKernel.KernelParameterMetadata},Microsoft.SemanticKernel.KernelReturnParameterMetadata,Microsoft.Extensions.Logging.ILoggerFactory)</Target>
+    <Left>lib/netstandard2.0/Microsoft.SemanticKernel.Core.dll</Left>
+    <Right>lib/netstandard2.0/Microsoft.SemanticKernel.Core.dll</Right>
+    <IsBaselineSuppression>true</IsBaselineSuppression>
+  </Suppression>
+</Suppressions>
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/CompatibilitySuppressions.xml:18]: Let's try to eliminate this also

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            "  }," +
+            "  \"servers\": [" +
+            "    {" +
+            "      \"url\": \"https://api.github.com\"" +
+            "    }" +
+            "  ]," +
+            "  \"paths\": {" +
+            "    \"/versions\": {" +
+            "      \"get\": {" +
+            "        \"x-openai-isConsequential\": false," +
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:50]: Make this true for the sake of the sample

File: dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs [hunk 10]
```diff
@@ -333,13 +346,24 @@ async Task<RestApiOperationResponse> ExecuteAsync(KernelArguments variables, Can
 
         var returnParameter = operation.GetDefaultReturnParameter();
 
+        // Add unstructured metadata, specific to Open API, to the metadata property bag.
+        var metadataPropertyBag = new Dictionary<string, object?>();
+        metadataPropertyBag.Add(OpenApiKernelExtensions.OperationExtensionsIdKey, operation.Id);
+        metadataPropertyBag.Add(OpenApiKernelExtensions.OperationExtensionsPathKey, operation.Path);
+        metadataPropertyBag.Add(OpenApiKernelExtensions.OperationExtensionsMethodKey, operation.Method.ToString());
+        if (operation.Extensions is not null && operation.Extensions.Any())
```
Comment (SergeyMenshykh) [inline dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs:354]:
```suggestion
        if (operation.Extensions is { Count: > 0 })
```
nit: the new pattern matching syntaxis that checks that the collection is initialized and contains elements.

File: dotnet/src/Functions/Functions.OpenApi/OpenApi/OpenApiDocumentParser.cs [hunk 11]
```diff
@@ -183,14 +186,63 @@ internal static List<RestApiOperation> CreateRestApiOperations(string? serverUrl
...
+                {
+                    var schemaBuilder = new StringBuilder();
+                    var jsonWriter = new OpenApiJsonWriter(new StringWriter(schemaBuilder, CultureInfo.InvariantCulture), new OpenApiJsonWriterSettings() { Terse = true });
+                    extension.Value.Write(jsonWriter, Microsoft.OpenApi.OpenApiSpecVersion.OpenApi3_0);
+                    extensionValueObj = schemaBuilder.ToString();
+                }
+            }
+            else
+            {
+                throw new KernelException($"The type of extension property '{extension.Key}' is not supported while trying to consume the OpenApi schema.");
```
Comment (SergeyMenshykh) [inline dotnet/src/Functions/Functions.OpenApi/OpenApi/OpenApiDocumentParser.cs:237]: I'm not sure if we need to throw an exception here. I'm considering cases when someone imports an OpenAPI specification that might have extensions like `OpenApiDeprecationExtension`, `OpenApiPagingExtension`, and other derivatives of `IOpenApiExtension`, but the scenario doesn't require those for execution. In those cases, we should not stop the import operation because we don't support the particular type of extension.  
   
On the other hand, there could be scenarios that require the extensions. To support them, they need to be reported, and corresponding functionality added.

File: dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs (hunk 9 above)
Comment (westey-m) [reply dotnet/samples/KernelSyntaxExamples/Example23_OpenAPIPlugin.cs:50]: I did consider it, but version is not a consequential action, so I felt making it true would be misleading.  Instead I suggested in the xml comments that people should play around with the example by changing the value and seeing the outcome.

File: dotnet/src/Functions/Functions.OpenApi/OpenApi/OpenApiDocumentParser.cs (hunk 11 above)
Comment (westey-m) [reply dotnet/src/Functions/Functions.OpenApi/OpenApi/OpenApiDocumentParser.cs:237]: Maybe the best is just to avoid adding them to the collection if we do not recognize them. That way we can always add them later, once we have an appropriate way of representing them.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs (hunk 6 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs]: Adopted `additionalMetadata` as suggested by @markwallace-microsoft

File: dotnet/src/SemanticKernel.Abstractions/CompatibilitySuppressions.xml (hunk 7 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/CompatibilitySuppressions.xml]: I used @dmytrostruk's suggestion to eliminate this one.  Will see if something similar works with the other one.

File: dotnet/src/SemanticKernel.Core/CompatibilitySuppressions.xml (hunk 8 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/CompatibilitySuppressions.xml]: Found a way to remove this breaking change, but it required me to introduce a new options parameter. I think this is better anyway, since there are too many optional parameters already, and this is a better pattern.

File: dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs (hunk 4 above)
Comment (westey-m) [reply dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs:214]: I'm looking into a few options to make these more easily accessible, and will get back on it later.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+using System.ComponentModel;
+using System.Reflection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Optional options that can be provided when creating a <see cref="KernelFunction"/> from a method.
+/// </summary>
+public sealed class KernelFunctionFromMethodOptions
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs:15]: Since it's `SemanticKernel.Core`, I'm wondering if we should mark this class as `Experimental` for the first period of time. Same for public places where this class is used. This will give us more space if we want to update it later, rename it or fix some issues. In this case we should remember to remove `Experimental` flag after some time.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    public string? FunctionName { get; init; }
+
+    /// <summary>
+    /// The description to use for the function. If null, it will default to one derived from the passed <see cref="Delegate"/> or <see cref="MethodInfo"/>, if possible
+    /// (e.g. via a <see cref="DescriptionAttribute"/> on the method).
+    /// </summary>
+    public string? Description { get; init; }
+
+    /// <summary>
+    /// >Optional parameter descriptions. If null, it will default to one derived from the passed <see cref="Delegate"/> or <see cref="MethodInfo"/>.
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs:29]:
```suggestion
    /// Optional parameter descriptions. If null, it will default to one derived from the passed <see cref="Delegate"/> or <see cref="MethodInfo"/>.
```

**Review by dmytrostruk — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs (hunk 12 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs:17]: Added, thanks.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs [hunk 14]
```diff
@@ -49,22 +50,50 @@ internal sealed class KernelFunctionFromMethod : KernelFunction
...
+
+    /// <summary>
+    /// Creates a <see cref="KernelFunction"/> instance for a method, specified via an <see cref="MethodInfo"/> instance
+    /// and an optional target object if the method is an instance method.
+    /// </summary>
+    /// <param name="method">The method to be represented via the created <see cref="KernelFunction"/>.</param>
+    /// <param name="target">The target object for the <paramref name="method"/> if it represents an instance method. This should be null if and only if <paramref name="method"/> is a static method.</param>
+    /// <param name="options">Optional function creation options.</param>
+    /// <returns>The created <see cref="KernelFunction"/> wrapper for <paramref name="method"/>.</returns>
+    [Experimental("SKEXP0001")]
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs:75]: This is an internal class so I don't think we need this

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+using System.Reflection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Optional options that can be provided when creating a <see cref="KernelFunction"/> from a method.
+/// </summary>
+[Experimental("SKEXP0001")]
+public sealed class KernelFunctionFromMethodOptions
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethodOptions.cs:17]: Does this need to be public?

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs [hunk 16]
```diff
@@ -49,22 +50,50 @@ internal sealed class KernelFunctionFromMethod : KernelFunction
...
+            method,
+            target,
+            new KernelFunctionFromMethodOptions
+            {
+                FunctionName = functionName,
+                Description = description,
+                Parameters = parameters,
+                ReturnParameter = returnParameter,
+                LoggerFactory = loggerFactory
+            });
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs:64]: Why is this change needed as part of this PR?

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs [hunk 17]
```diff
@@ -35,6 +35,18 @@ public static class KernelFunctionFactory
         ILoggerFactory? loggerFactory = null) =>
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs:35]: Not for this PR, but do we want to mark this old method as obsolete (at some time in the future) to encourage developers to use the new method?

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs (hunk 14 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs]: Oops, good catch.  Removed.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs (hunk 17 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFactory.cs:35]: Agreed, I also think we should obsolete the existing ones, and makes sense to do it in a separate PR.

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs [hunk 18]
```diff
@@ -117,18 +117,19 @@ internal KernelFunction(string name, string description, IReadOnlyList<KernelPar
...
         Verify.ParametersUniqueness(parameters);
 
-        this.Metadata = new KernelFunctionMetadata(name, properties)
+        this.Metadata = new KernelFunctionMetadata(name)
         {
             PluginName = pluginName,
             Description = description,
             Parameters = parameters,
             ReturnParameter = returnParameter ?? KernelReturnParameterMetadata.Empty,
+            AdditionalProperties = additionalMetadata ?? KernelFunctionMetadata.s_emptyDictionary,
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs:132]: Should this be named `AdditionalMetadata`?

Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs:132]: I think outside of the metadata class AdditionalMetadata makes sense, since it's clarifying where the data will end up.  However inside the metadata class the word metadata is unnecessary and properties make more sense.

File: dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs [hunk 19]
```diff
@@ -206,7 +206,7 @@ private record struct MethodDetails(string Name, string Description, Implementat
         IReadOnlyList<KernelParameterMetadata> parameters,
         KernelReturnParameterMetadata returnParameter,
         ReadOnlyDictionary<string, object?>? additionalMetadata = null) :
-        base(functionName, pluginName, description, parameters, returnParameter, properties: additionalMetadata)
+        base(functionName, pluginName, description, parameters, returnParameter, additionalMetadata: additionalMetadata)
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs:209]: I'd keep the naming consistent either `additionalMetadata` or `additionalProperties` everywhere

Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromMethod.cs:209]: I think context matters.  Outside of the metadata class AdditionalMetadata makes sense, since it's clarifying where the data will end up. However inside the metadata class the word metadata is unnecessary and properties make more sense.

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs [hunk 20]
```diff
@@ -91,4 +97,15 @@ public KernelReturnParameterMetadata ReturnParameter
             this._returnParameter = value;
         }
     }
+
+    /// <summary>Gets optional metadata in addition to the named properties already available on this class.</summary>
+    public ReadOnlyDictionary<string, object?> AdditionalProperties
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs:102]: Add the experimental marker

Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionMetadata.cs:103]: Makes sense.  Done.

**Review by SergeyMenshykh — APPROVED** [review]
